### PR TITLE
Implement User flow

### PR DESCRIPTION
**Pull request by Mvamd12**
Overview:
This PR introduces the user model and the user route for the application. The primary focus is on creating a user model that encapsulates user attributes and implementing an endpoint to create a new user with password validation.

Notes:
The other endpoints will be worked on after setting up the authentication mechanism.

Next Step :
implementing authentication and user login features.

### CODE REVIEW

**Comment by Omarmanassrah** [comment]
I`m approved for mohammad request

**Comment by Mohammed-samaheen** [comment]
> I`m approved for mohammad request

that is not how we approve, you should go to the file changes and then add your review

File: app/api/routes/user.py [hunk 1]
```diff
@@ -1,8 +1,41 @@
...
+
+#Password regex
+password_regex = (
+    r'^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[@#$%^&+=]).{8,}$'
+)
+
+@router.post("/users/", response_model=User, status_code=status.HTTP_201_CREATED)
+async def create_user(user: UserCreate):
+    #Validate password
+    if not re.match(password_regex, user.password):
```
Comment (Mohammed-samaheen) [inline app/api/routes/user.py]: any input validation should be in the validation model, in UserCreateRequestModel

File: app/models.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from uuid import UUID, uuid4
+from datetime import datetime, timezone
+from pydantic import BaseModel, EmailStr, Field
+from passlib.context import CryptContext
+
+#Set up password hashing context using bcrypt
+pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
+
+class UserBase(BaseModel):
```
Comment (Mohammed-samaheen) [inline app/models.py]: UserBaseModel

File: app/models.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    username: str
+    email: EmailStr
+    is_admin: bool = False
+    is_active: bool = True
+
+class UserCreate(UserBase):
```
Comment (Mohammed-samaheen) [inline app/models.py]: UserCreateRequestModel

File: app/models.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    password: str
+
+#Inheriting from UserBase
+class User(UserBase):
+    id: UUID = Field(default_factory=uuid4)
+    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
+    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
```
Comment (Mohammed-samaheen) [inline app/models.py]: default value for the update date should be None (NULL)

File: app/models.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    def get_password_hash(self):
+        return pwd_context.hash(self.password)
+    
+    #Method to verify if the provided password matches the hashed password
+    def verify_password(self, password: str):
+        return pwd_context.verify(password, self.get_password_hash())
+    
+    def to_dict(self):
```
Comment (Mohammed-samaheen) [inline app/models.py]: There is no need for this one, we can use model_dump form BaseModel

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/models.py (hunk 5 above)
Comment (Mvamd12) [reply app/models.py]: done.

File: app/models.py (hunk 4 above)
Comment (Mvamd12) [reply app/models.py]: done.

File: app/models.py (hunk 3 above)
Comment (Mvamd12) [reply app/models.py]: done.

File: app/models.py (hunk 2 above)
Comment (Mvamd12) [reply app/models.py]: done.

File: app/api/routes/user.py (hunk 1 above)
Comment (Mvamd12) [reply app/api/routes/user.py]: I replace it by validator in UserCreateRequestModel, done.

File: app/api/routes/user.py [hunk 6]
```diff
@@ -1,8 +1,29 @@
...
 
-# make sure to remove this function
-@router.get("/example")
-def example():
-    return {"message": "this is an example"}
+
+users_db = {}
+
+
+@router.post("/users/", response_model=User, status_code=status.HTTP_201_CREATED)
```
Comment (Mohammed-samaheen) [inline app/api/routes/user.py]: Why was the UserCreateResponseModel removed? Password shouldn't be part of the response, and the same goes for the update date.

File: app/models.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+class UserBaseModel(BaseModel):
+    username: str
+    email: EmailStr
+    is_admin: bool = False
+    is_active: bool = True
+
+class UserCreateRequestModel(UserBaseModel):
+    password: str
+    @field_validator('password')
```
Comment (Mohammed-samaheen) [inline app/models.py]: There is no need for this complexity you can add it like this
```
password: str = Field(
        ...,
        min_length=8,
        pattern= password_regex,
    )
```

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/api/routes/user.py (hunk 6 above)
Comment (Mvamd12) [reply app/api/routes/user.py]: I fixed it, sorry i didn't realize it then.

File: app/models.py (hunk 7 above)
Comment (Mvamd12) [reply app/models.py]: i fixed it also, some mistakes happen to me while switching between branches, i was not focused in something.
Thank you.

File: app/api/routes/user.py (hunk 6 above)
Comment (Mohammed-samaheen) [reply app/api/routes/user.py]: UserCreateResponseModel should include the creation date. Please double-check the task to ensure the correct response format.

File: app/models.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#Set up password hashing context using bcrypt
+pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
+
+
+password_regex = r"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&#])[A-Za-z\d@$!%*?&#]{8,}$"
+
+class UserBaseModel(BaseModel):
+    username: str
+    email: EmailStr
+    is_admin: bool = False
```
Comment (Mohammed-samaheen) [inline app/models.py]: UserCreateRequestModel shouldn't have either is_admin or is_active. Please double-check the task.

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/api/routes/user.py (hunk 6 above)
Comment (Mvamd12) [reply app/api/routes/user.py]: yes iam just missing it, it's done.

File: app/models.py (hunk 8 above)
Comment (Mvamd12) [reply app/models.py]: done.

Comment (Mohammed-samaheen) [reply app/models.py]: UserBaseModel should have only username and email.
UserCreateRequestModel and UserCreateResponseModel should utilize UserBaseModel and add the necessary fields to it.

Comment (Mvamd12) [reply app/models.py]: I hope last code update meet the task models requirements, please dont worry about password verfication and security things i'm handling it in the second PR with fares.

File: app/models.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        pattern=password_regex,
+        description="Password must be at least 8 characters long, contain at least 1 uppercase letter, 1 lowercase letter, 1 digit, and 1 special character."
+    )
+
+class UserCreateResponseModel(UserBaseModel):
+    id: UUID
+    is_admin: bool = Field(default=False, description="Admin privileges.")
+    is_active: bool = Field(default=True, description="Account active status.")
+    created_at: datetime
+    updated_at: datetime = None
```
Comment (Mohammed-samaheen) [inline app/models.py]: Why is there an update date for the UserCreateResponseModel?

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

File: app/models.py (hunk 9 above)
Comment (Mvamd12) [reply app/models.py]: I'am added updated date again it's my fault, i fix it

**Review by Mohammed-samaheen — APPROVED** [review]
LGTM

**Pull request merged**